src/index.ts

```diff
@@ -43,19 +64,37 @@ export class Amplitude {
    * (whichever comes first), as well as on app close.
    *
    * @param eventType The name of the event you wish to track.
+   * @param eventProperties The event's properties.
+   * @param extra Extra untyped parameters for use in middleware.
    */
   logEvent(
     eventType: string,
     eventProperties?: Record<string, unknown>,
+    extra?: MiddlewareExtra,
   ): Promise<boolean> {
-    if (eventProperties && Object.keys(eventProperties).length > 0) {
-      return AmplitudeReactNative.logEventWithProperties(
-        this.instanceName,
-        eventType,
-        eventProperties,
-      );
-    }
-    return AmplitudeReactNative.logEvent(this.instanceName, eventType);
+    return AmplitudeReactNative.getUserId(this.instanceName).then((userId) =>
+      AmplitudeReactNative.getDeviceId(this.instanceName).then((deviceId) => {
+        const properties = eventProperties ?? {};
+        const event = {
+          event_type: eventType,
+          user_id: userId,
+          device_id: deviceId,
+          event_properties: properties,
+        };
+        if (!this._runMiddlewares(event, extra)) {
+          return Promise.resolve(false);
+        }
+
+        if (Object.keys(properties).length > 0) {
+          return AmplitudeReactNative.logEventWithProperties(
+            this.instanceName,
+            eventType,
+            properties,
+          );
+        }
+        return AmplitudeReactNative.logEvent(this.instanceName, eventType);
+      }),
+    );
```

Comment:
Ampli SDK has method `identify(userId: string | undefined, properties: IdentifyProperties, options?: EventOptions, extra?: MiddlewareExtra)`. If the `userId` is set (the `userId` argument, the `options` argument), `this.amplitude?.setUserId` is called before middleware/main logic.
How can the `segment` middleware get the userId if it wants to use it to call own `identify` method?
The `Amplitude-ReactNative` has method `identify(identifyInstance: Identify,  extra?: MiddlewareExtra)` and should set `event.user_id` for middlewares - here we can pass the `userId` in the `extra` argument, add the new `userId` argument or use `AmplitudeReactNative.getUserId()`.

Reply:
All native wrappers return promises (`getUserId`, `setUserId`, etc). Can we switch to `async`/`await` style to avoid nested `.then()`?

Reply:
I think it should be ok to use async/await. Can you please test it though.

Reply:
Given the current state of the SDKs I agree there isn't a great way to pass the `userId`. For now I think it can be a known limitation that the end user will need to manually retrieve the user/device id if needed in middleware. 

I am concerned adding too much logic to support middleware could impact performance. This should be resolved in our next-gen SDKs (in progress) that will have uniform event/method interfaces that will provide access to userId. 

cc @qingzhuozhen

Reply:
Fixed